core/rio/hdt/src/main/java/org/eclipse/rdf4j/rio/hdt/HDTArrayLog64.java

```diff
@@ -26,20 +29,32 @@
  * ...+---------+-------+
  * </pre>
  * 
- * Entries are stored little-endian, with each entry using <code>nrbits</code> bits
+ * Entries are stored little-endian, with each entry using <code>nrbits</code> bits, which is the log2(max_value). E.g.
+ * to store a maximum value of 1024 only 10 bits are needed, instead of storing a 16-bit <code>short</code>.
  * 
  * @author Bart Hanssens
  */
 class HDTArrayLog64 extends HDTArray {
 	private byte buffer[];
 
+	@Override
+	protected void setMaxValue(int maxval) {
+		// ceil(log2(maxval))
+		int i = 0;
+		while (++i < 32 && (maxval >> i) > 0)
+			;
+		setNrBits(i);
```

Comment:
I'm not sure what this loop does, but it looks like it both increments `i` and does some bitshifting. Could this be written more explicitly.

Reply:
Basically it checks the position of the highest bit set counting from 1

E.g. for storing value 7, or `0000 0000 0000 0111` b, only 3 bits are needed

Or ceil(log2(value)), hence the original HDT name "arraylog64" I guess, with a max of 64 bits in the c version)

Reply:
That would be nice as a comment in the code. I could google things based on what you wrote now and understand the code a lot better :)

Reply:
Btw. Could this method help you in any way? https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#highestOneBit-int- 

But maybe not if it works on two-complement....

Reply:
Or maybe: `ceil(log2(x)) = 32 - Integer.numberOfLeadingZeros(x - 1)`

https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#numberOfLeadingZeros-int-

Reply:
Indeed, the latter works as expected, thanks